lib/app_profiler/yarn/command.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+module AppProfiler
+  module Yarn
+    module Command
+      class YarnError < StandardError; end
+
+      mattr_accessor :yarn_setup, default: false
+
+      def yarn(command, *options)
+        setup_yarn unless yarn_setup
+        ensure_command_valid(command)
+        exec("yarn", command, *options) do
+          raise YarnError, "Failed to run #{command}."
+        end
+      end
+
+      def setup_yarn
+        ensure_yarn_installed
+        yarn("init", "--yes") unless package_json_exists?
+      end
+
+      private
+
+      def ensure_command_valid(command)
+        unless command.in?(%w(init add run))
+          raise YarnError, "Illegal command #{command}."
+        end
+      end
+
+      def ensure_yarn_installed
+        exec("which yarn > /dev/null") do
+          raise(
+            YarnError,
+            <<~MSG.squish
+              `yarn` command not found.
+              Please install `yarn` or make it available in PATH.
+            MSG
+          )
+        end
+        self.yarn_setup = true
+      end
+
+      def package_json_exists?
+        AppProfiler.root.join("package.json").exist?
+      end
+
+      def exec(*command)
+        system(*command).tap do |return_code|
+          yield unless return_code
+        end
+      end
```

Comment:
Since we would never want to run `exec` without calling `ensure_command_valid?` can we move the call to that function here? That way down the road if `exec` gets reused somewhere, we can trust that it will always have its safeguards for checking the command bundled into it.

Reply:
I changed it to permit all command strings. This should be the most secure option despite the cost of iteration and matching.

Reply:
> hardcode these strings into the methods and forgo allow-listing

Can you expand on what you mean by this? What you've done here is fantastic with the allow list, if we can maintain a similar strict list of commands another way i'm ok with that as well 👍